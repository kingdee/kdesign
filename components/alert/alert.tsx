import React, { ReactNode, useContext, useEffect, useState, useRef, ReactElement, useCallback } from 'react'
import { Icon } from '../index'
import ReactDOM from 'react-dom'
import { tuple } from '../_utils/type'
import classNames from 'classnames'
import ConfigContext from '../config-provider/ConfigContext'
import { getCompProps } from '../_utils'
import { usePrevious } from '../_utils/hooks'
import devWarning from '../_utils/devwarning'

export const AlertTypes = tuple('success', 'info', 'warning', 'error')
export type AlertType = typeof AlertTypes[number]
const IconTypes = tuple('right-solid', 'notice', 'warning-solid', 'warning-solid')
type IconType = typeof IconTypes[number]
type Destination = 'bg' | 'border' | 'icon'
type BannerOffsetType = [number, number]
export const allBannerAlertPortals: Map<string, HTMLElement> = new Map()
const allBannerAlertPortalsContainer = document.querySelector('body')!
// 由于是全局的, 暂时关闭
// let originOrder = 100
export interface IAlertProps {
  type?: AlertType
  message?: ReactNode
  showIcon?: boolean
  icon?: ReactNode
  closable?: boolean
  closeNode?: ReactNode
  delayOffTime?: number
  className?: string
  style?: Record<string, unknown>
  prefixCls?: string // 样式前缀
  onClose?: () => void
  extra?: ReactNode
  banner?: boolean
  bannerOffset?: BannerOffsetType
}
const setContainerStyle = (dom: HTMLElement, posInfo: BannerOffsetType) => {
  const styles: Record<string, string> = {
    left: posInfo[0] + 'px',
    top: posInfo[1] + 'px',
  }
  Object.keys(styles).forEach((prop) => {
    dom.style.setProperty(prop, styles[prop])
  })
}

/**
 * @description 创建悬浮顶部的反馈浮层容器
 * @param alertContainer 反馈浮层容器ref
 * @param alertPrefixCls 反馈浮层前缀class
 * @param bannerOffset 反馈浮层悬浮顶部的坐标量
 *
 */

function createAlertContainer(
  alertContainer: React.MutableRefObject<HTMLElement | null>,
  alertPrefixCls: string,
  bannerOffset: BannerOffsetType,
) {
  const posKey = bannerOffset!.join(',')
  const container = document.createElement('div')
  container.classList.add(`${alertPrefixCls}-banner-container`)
  // container.classList.add(`${alertPrefixCls}-banner-container-flex`)

  setContainerStyle(container, bannerOffset!)
  alertContainer.current = container
  allBannerAlertPortalsContainer.appendChild(container)
  allBannerAlertPortals.set(posKey, container)
}
const Alert: React.FC<IAlertProps> = (props: IAlertProps): ReactElement | React.ReactPortal => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps } = useContext(ConfigContext)
  const alertProps: IAlertProps = getCompProps('Alert', userDefaultProps, props) // 属性需要合并一遍用户定义的默认属性
  const {
    type,
    message,
    showIcon,
    icon,
    className,
    style,
    prefixCls: customPrefixcls,
    closable,
    closeNode,
    delayOffTime,
    onClose,
    extra,
    banner,
    bannerOffset,
    ...othersProps
  } = alertProps
  devWarning(AlertTypes.indexOf(type!) === -1, 'alert', `cannot found alert type '${type}'`)
  const [visible, setVisible] = useState(true)
  const previousBannerOffset = usePrevious<BannerOffsetType>(bannerOffset!)
  const timerRef = useRef<number | undefined>()
  const alertRef = useRef<HTMLDivElement>(null)
  const alertContainer = useRef<HTMLElement | null>(allBannerAlertPortalsContainer)
  const forceUpdate = useState(0)[1]
  const previousMessage = usePrevious(message)
  const alertPrefixCls = getPrefixCls!(prefixCls, 'alert', customPrefixcls)

  useEffect(() => {
    if (banner) {
      const previousPosKey = previousBannerOffset.join(',')
      const posKey = bannerOffset!.join(',')
      if (
        previousPosKey !== posKey ||
        !alertContainer.current ||
        alertContainer.current === allBannerAlertPortalsContainer
      ) {
        const oldContainer = allBannerAlertPortals.get(previousPosKey)
        setTimeout(() => {
          if (oldContainer?.children?.length === 0) {
            Array.from(allBannerAlertPortalsContainer.children).includes(oldContainer) &&
              allBannerAlertPortalsContainer.removeChild(oldContainer)
            allBannerAlertPortals.delete(previousPosKey)
          }
        }, 0)
        const newContainer = allBannerAlertPortals.get(posKey)
        if (newContainer) {
          alertContainer.current = newContainer
        } else {
          createAlertContainer(alertContainer, alertPrefixCls, bannerOffset!)
        }
      }
      forceUpdate((val) => val + 1)
    }
  }, [banner, bannerOffset, previousBannerOffset, forceUpdate, alertPrefixCls])

  const showAlert = () => {
    setVisible(true)
  }

  const hideAlert = useCallback(() => {
    timerRef.current = window.setTimeout(() => {
      setVisible(false)
    }, delayOffTime)
  }, [delayOffTime])

  useEffect(() => {
    if (delayOffTime === 0 || message !== previousMessage) {
      showAlert()
      clearTimer()
    }
    if (delayOffTime === 0) {
      return () => {
        clearTimeout(timerRef.current)
      }
    }
    hideAlert()
    return () => {
      clearTimeout(timerRef.current)
    }
  }, [delayOffTime, message, previousMessage, timerRef, hideAlert])

  const alertClasses = classNames(alertPrefixCls, className, {
    [`${alertPrefixCls}-closable`]: closable,
    [`${alertPrefixCls}-banner`]: banner,
    [`${alertPrefixCls}-leave`]: !visible || !message,
    [`${alertPrefixCls}-visible`]: visible && message,
    [`${alertPrefixCls}-container`]: true,
  })

  const handleClose = () => {
    onClose && onClose()
    setVisible(false)
    clearTimer()
  }
  function clearTimer() {
    if (timerRef.current) {
      clearTimeout(timerRef.current)
      timerRef.current = undefined
    }
  }
  function handleMouseOut() {
    hideAlert()
  }
  const events: Record<string, () => void> = {}
  if (delayOffTime) {
    events.onMouseOver = clearTimer
    events.onMouseOut = handleMouseOut
  }

  const alertIconType: IconType = IconTypes[AlertTypes.indexOf(type!)]
  const getAlertColorType = (mode: Destination): string => {
    return `alert-${type}-${mode}-color`
  }
  const iconNode: ReactNode = (
    <div className={classNames(`${alertPrefixCls}-icon`, getAlertColorType('icon'))}>
      {icon || <Icon type={alertIconType} />}
    </div>
  )
  const AlertComp: React.ReactElement = (
    <div
      className={classNames(alertClasses, getAlertColorType('bg'), getAlertColorType('border'))}
      {...events}
      ref={alertRef}
      style={style}
      {...othersProps}
    >
      {showIcon && iconNode}
      <div
        className={classNames({
          [`${alertPrefixCls}-message`]: true,
          [`${alertPrefixCls}-without-icon-message`]: !showIcon,
        })}
      >
        {message}
      </div>
      {extra}
      {closable && type !== 'success' && (
        <div className={`${alertPrefixCls}-close-icon`} onClick={handleClose}>
          {closeNode || <Icon type="close" />}
        </div>
      )}
    </div>
  )
  if (banner) {
    return ReactDOM.createPortal(AlertComp, alertContainer.current!)
  } else {
    return AlertComp
  }
}

Alert.displayName = 'Alert'
export default Alert
