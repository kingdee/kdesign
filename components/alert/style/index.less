@import '../../style/themes/index';
@import './mixin.less';
@import './token.less';

@alert-prefix-cls: ~'@{kd-prefix}-alert';

@success: success;
@warning: warning;
@error: error;
@info: info;

.alert-bg-type-set() {
  @success: @alert-success-bg-color;
  @warning: @alert-warning-bg-color;
  @error: @alert-error-bg-color;
  @info: @alert-info-bg-color;
}

.alert-icon-type-set() {
  @success: @alert-success-icon-color;
  @warning: @alert-warning-icon-color;
  @error: @alert-error-icon-color;
  @info: @alert-info-icon-color;
}

.alert-border-type-set() {
  @success: @alert-success-border-color;
  @warning: @alert-warning-border-color;
  @error: @alert-error-border-color;
  @info: @alert-info-border-color;
}

@alert-transition-fn: cubic-bezier(0.32, 0.94, 0.6, 1);

.icon-center {
  i {
    align-self: center;
  }
}

.@{alert-prefix-cls} {
  &-container {
    z-index: @z-index-apex;
    display: none;
    padding: @alert-padding-vertical @alert-padding-horizontal;
    align-items: center;
    transition: display 3s;
    border-radius: @alert-radius-border;
    border: 1px solid;
  }
  &-banner-container {
    display: flex;
    flex-direction: column;
    position: fixed;
    z-index: @z-index-apex;
    width: 100%;
  }
  each(.alert-bg-type-set(), {
    .alert-bg-type(@key, @value);
  });
  each(.alert-border-type-set(), {
    .alert-border-type(@key, @value);
  });
  &-icon {
    display: flex;
    cursor: default;
    .icon-center();
    height: @alert-message-icon-font-size;
    margin-right: @alert-message-icon-margin-right;
    font-size: @alert-message-icon-font-size;
    each(.alert-icon-type-set(), {
      .alert-icon-type(@key, @value);
    });
  }
  &-message {
    color: @alert-message-color-text;
    font-size: @alert-message-font-size;
    line-height: 16px;
    display: flex;
    flex-grow: 1;
    cursor: default;
  }
  &-leave {
    transition: min-height @duration-promptly @alert-transition-fn, opacity @duration-promptly @alert-transition-fn,
      margin @duration-promptly @alert-transition-fn, height @duration-promptly @alert-transition-fn,
      line-height @duration-promptly @alert-transition-fn;
    display: flex;
    min-height: 0;
    margin: 0 !important;
    overflow: hidden;
    opacity: 0;

    .@{alert-prefix-cls}-icon,
    .@{alert-prefix-cls}-message,
    .@{alert-prefix-cls}-close-icon {
      height: 0;
      line-height: 0;
    }
  }
  &-close-icon {
    display: flex;
    height: @alert-message-icon-font-size;
    margin-left: @alert-close-icon-margin-left;
    color: @alert-close-icon-color;
    font-size: @alert-close-icon-font-size;
    cursor: pointer;
    &:hover {
      color: @alert-close-icon-color-hover;
    }
    &:active {
      color: @color-theme-click;
    }
    .icon-center();
  }
  &-visible {
    display: flex;
  }
  &-banner {
    position: fixed;
    animation: moveIn @duration-promptly @alert-transition-fn both;
  }
}
@keyframes moveIn {
  from {
    transform: translateY(-50px);
  }
  to {
    transform: translateY(0);
  }
}
