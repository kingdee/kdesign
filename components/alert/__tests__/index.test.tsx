import React from 'react'
import { mount, render, ReactWrapper } from 'enzyme'
import Alert from '../index'
import { AlertTypes, AlertType, allBannerAlertPortals } from '../alert'
import { act } from 'react-dom/test-utils'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'
jest.useFakeTimers()

describe('Alert', () => {
  // 1.mount test
  mountTest(Alert)
  AlertTypes.forEach((type) => {
    mountTest(() => <Alert type={type} delayOffTime={0} />)
  })

  // 2.render test
  it('renders correctly', () => {
    expect(render(<Alert />)).toMatchSnapshot()
    AlertTypes.forEach((type) => {
      expect(render(<Alert type={type} delayOffTime={0} />)).toMatchSnapshot()
    })
  })
  // 3. warns in components
  it('warns if type is wrong', () => {
    const mockWarn = jest.fn()
    jest.spyOn(console, 'warn').mockImplementation(mockWarn)
    const type = 'who am I' as any as AlertType
    render(<Alert type={type} delayOffTime={0} />)
    expect(mockWarn).toHaveBeenCalledTimes(1)
    expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-alert: cannot found alert type 'who am I'")
  })

  // 4. render null or undefined without errors
  it('render null or undefined without errors', () => {
    const wrapper = mount(
      <Alert>
        {null}
        {undefined}
      </Alert>,
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.kd-alert-message').text()).toBe('')
  })

  // 5. displayName
  it('should have displayName static property', () => {
    const wrapper = mount(<Alert type={'success'} delayOffTime={0}></Alert>)
    expect((wrapper.type() as any).displayName).toBe('Alert')
  })

  // 6.API Test
  describe('API test', () => {
    it('should delayOffTime, is zero, can be stay at original placement; is not zero, can be unvisible after delayOffTime ms', async () => {
      let wrapper: ReactWrapper
      await act(async () => {
        wrapper = mount(<Alert type={'success'} message={'Alert Component1'} delayOffTime={0} />)
      })
      let unvisibleWrapper: ReactWrapper
      await act(async () => {
        unvisibleWrapper = mount(<Alert type={'success'} message={'Alert Component1'} delayOffTime={1000} />)
      })
      expect(wrapper!.exists('.kd-alert-visible')).toBe(true)
      jest.runTimersToTime(1000)
      unvisibleWrapper!.update()
      expect(wrapper!.exists('.kd-alert-visible')).toBe(true)
      expect(unvisibleWrapper!.exists('.kd-alert-visible')).toBe(false)
      unvisibleWrapper!.setProps({
        delayOffTime: 0,
      })
      unvisibleWrapper!.update()
      expect(unvisibleWrapper!.exists('.kd-alert-visible')).toBe(true)
      wrapper!.setProps({
        delayOffTime: 1000,
      })
      expect(wrapper!.exists('.kd-alert-visible')).toBe(true)
      jest.runTimersToTime(1000)
      wrapper!.update()
      expect(wrapper!.exists('.kd-alert-visible')).toBe(false)
    })

    it('should showIcon & closable use right', () => {
      const showIconClosableArray = [
        [true, true],
        [true, false],
        [false, true],
        [false, false],
      ]
      showIconClosableArray.forEach(([showIcon, closable]) => {
        expect(
          render(<Alert type={'success'} showIcon={showIcon} closable={closable} delayOffTime={0} />),
        ).toMatchSnapshot()
      })
    })

    it('should class use right', () => {
      const wrapper = mount(
        <Alert type={'success'} message={'Alert Component'} delayOffTime={0} className={'class-test'} />,
      )
      expect(wrapper.exists('.class-test')).toBe(true)
    })

    it('should banner & bannerOffset use right', () => {
      const wrapper = mount(
        <Alert type={'success'} message={'Alert Component'} delayOffTime={0} banner className={'banner-test-a'} />,
      )
      const containerleft0Top0 = document.body.querySelector('.kd-alert-banner-container')
      expect(getComputedStyle(containerleft0Top0!, null).getPropertyValue('left')).toBe('0px')
      expect(getComputedStyle(containerleft0Top0!, null).getPropertyValue('top')).toBe('0px')
      expect(containerleft0Top0?.children.length).toBe(1)
      expect(allBannerAlertPortals.size).toBe(1)
      const otherWrapper = mount(
        <Alert type={'success'} message={'Alert Component1'} delayOffTime={0} banner className={'banner-test-b'} />,
      )
      expect(containerleft0Top0?.children.length).toBe(2)
      expect(allBannerAlertPortals.size).toBe(1)
      expect(containerleft0Top0?.children[0].classList.contains('banner-test-a')).toBe(true)
      expect(containerleft0Top0?.children[1].classList.contains('banner-test-b')).toBe(true)
      otherWrapper.setProps({
        bannerOffset: [10, 10],
      })
      otherWrapper.update()
      expect(document.body.querySelectorAll('.kd-alert-banner-container').length).toBe(2)
      expect(allBannerAlertPortals.size).toBe(2)
      const containerleft10Top10 = document.body.querySelectorAll('.kd-alert-banner-container')[1]
      expect(containerleft10Top10?.children[0].classList.contains('banner-test-b')).toBe(true)
      expect(containerleft0Top0?.children.length).toBe(1)
      expect(containerleft10Top10?.children.length).toBe(1)
      expect(getComputedStyle(containerleft10Top10!, null).getPropertyValue('left')).toBe('10px')
      expect(getComputedStyle(containerleft10Top10!, null).getPropertyValue('top')).toBe('10px')
      wrapper.setProps({
        bannerOffset: [10, 20],
      })
      wrapper.update()
      // containerleft0Top0 to be removed
      jest.runAllTimers()
      const containerleft10Top20 = document.body.querySelectorAll('.kd-alert-banner-container')[1]
      expect(containerleft10Top20?.children[0].classList.contains('banner-test-a')).toBe(true)
      expect(containerleft10Top10?.children.length).toBe(1)
      expect(containerleft10Top20?.children.length).toBe(1)
      expect(containerleft0Top0?.children.length).toBe(0)
      expect(document.body.querySelectorAll('.kd-alert-banner-container').length).toBe(2)
      expect(allBannerAlertPortals.size).toBe(2)
      expect(getComputedStyle(containerleft10Top10!, null).getPropertyValue('left')).toBe('10px')
      expect(getComputedStyle(containerleft10Top10!, null).getPropertyValue('top')).toBe('10px')
      expect(getComputedStyle(containerleft10Top20!, null).getPropertyValue('left')).toBe('10px')
      expect(getComputedStyle(containerleft10Top20!, null).getPropertyValue('top')).toBe('20px')
    })

    it('should icon use right', () => {
      const iconFC = function () {
        return <div className="icon-customed" />
      }
      const wrapper = mount(
        <Alert type={'success'} message={'Alert Component'} delayOffTime={0} banner icon={iconFC()} />,
      )
      expect(wrapper.exists('.icon-customed')).toBe(false)
      wrapper.setProps({
        showIcon: true,
      })
      wrapper.update()
      expect(wrapper.exists('.icon-customed')).toBe(true)
    })

    it('should closeNode use right', () => {
      const closeNodeFC = function () {
        return <div className="closeNode-customed" />
      }
      const wrapper = mount(
        <Alert type={'info'} message={'Alert Component'} delayOffTime={0} banner closeNode={closeNodeFC()} />,
      )
      expect(wrapper.exists('.closeNode-customed')).toBe(false)
      wrapper.setProps({
        closable: true,
      })
      wrapper.update()
      expect(wrapper.exists('.closeNode-customed')).toBe(true)
    })

    it('should extra use right', () => {
      const extraFC = function () {
        return <div className="extra-customed" />
      }
      const wrapper = mount(
        <Alert type={'success'} message={'Alert Component'} delayOffTime={0} banner extra={extraFC()} />,
      )
      expect(wrapper.exists('.extra-customed')).toBe(true)
    })

    it('test click', () => {
      const onClose = jest.fn()
      const wrapper = mount(<Alert type={'info'} showIcon closable onClose={onClose} delayOffTime={0} />)
      act(() => {
        wrapper.find('.kd-alert-close-icon').simulate('click')
      })
      expect(onClose).toHaveBeenCalled()
    })

    it('test mouseover & mouseout', async () => {
      const onClose = jest.fn()
      let wrapper: ReactWrapper
      await act(async () => {
        wrapper = mount(
          <Alert type={'success'} message={'test mouseover & mouseout'} showIcon closable onClose={onClose} />,
        )
      })
      jest.runTimersToTime(4900)
      expect(wrapper!.exists('.kd-alert-visible')).toBe(true)
      await act(async () => {
        wrapper!.find('.kd-alert-container').simulate('mouseover')
      })
      jest.runTimersToTime(5000)
      wrapper!.update()
      expect(wrapper!.exists('.kd-alert-visible')).toBe(true)
      await act(async () => {
        wrapper.find('.kd-alert-container').simulate('mouseout')
      })
      jest.runTimersToTime(5000)
      wrapper!.update()
      expect(wrapper!.exists('.kd-alert-visible')).toBe(false)
      expect(onClose).not.toHaveBeenCalled()
    })
  })
  // 7. class or style state
  it('should className style useful', () => {
    const wrapper = mount(<Alert className="test-alert" data-test="test" style={{ margin: '1px' }}></Alert>)
    expect(wrapper.find('.kd-alert-container')).toHaveClassName('test-alert')
    expect(wrapper.find('.kd-alert-container')).toHaveStyle({ margin: '1px' })
    expect(wrapper.find('.kd-alert-container').prop('data-test')).toEqual('test')
  })
  // 8. component interaction(event)
  // 9. config Provider
  it('should config use config provider', () => {
    const alertConfig = {
      compDefaultProps: {
        Alert: {
          type: 'success',
        },
      },
    }
    const wrapper = mount(
      <ConfigProvider value={alertConfig}>
        <Alert delayOffTime={0}></Alert>
      </ConfigProvider>,
    )
    expect(wrapper.find('.kd-alert-container')).toHaveClassName('.alert-success-bg-color')
  })

  // 10. ref test
  // 11. special case
})
