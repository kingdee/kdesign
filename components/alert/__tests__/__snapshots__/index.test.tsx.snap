// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert API test should showIcon & closable use right 1`] = `
<div
  class="kd-alert kd-alert-closable kd-alert-leave kd-alert-container alert-success-bg-color alert-success-border-color"
>
  <div
    class="kd-alert-icon alert-success-icon-color"
  >
    <i
      class="kdicon kdicon-right-solid"
    />
  </div>
  <div
    class="kd-alert-message"
  />
</div>
`;

exports[`Alert API test should showIcon & closable use right 2`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-success-bg-color alert-success-border-color"
>
  <div
    class="kd-alert-icon alert-success-icon-color"
  >
    <i
      class="kdicon kdicon-right-solid"
    />
  </div>
  <div
    class="kd-alert-message"
  />
</div>
`;

exports[`Alert API test should showIcon & closable use right 3`] = `
<div
  class="kd-alert kd-alert-closable kd-alert-leave kd-alert-container alert-success-bg-color alert-success-border-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert API test should showIcon & closable use right 4`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-success-bg-color alert-success-border-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert render null or undefined without errors 1`] = `
<Alert>
  <div
    className="kd-alert kd-alert-leave kd-alert-container alert-warning-bg-color alert-warning-border-color"
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="kd-alert-message kd-alert-without-icon-message"
    />
  </div>
</Alert>
`;

exports[`Alert renders correctly 1`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-warning-bg-color alert-warning-border-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert renders correctly 2`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-success-bg-color alert-success-border-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert renders correctly 3`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-info-bg-color alert-info-border-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert renders correctly 4`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-warning-bg-color alert-warning-border-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;

exports[`Alert renders correctly 5`] = `
<div
  class="kd-alert kd-alert-leave kd-alert-container alert-error-bg-color alert-error-border-color"
>
  <div
    class="kd-alert-message kd-alert-without-icon-message"
  />
</div>
`;
